File: fbeuler_gs.m
function [ x,y,t,m ,convInd] = fbeuler_gs( dx, dy, x0, T, N, tol, maxiter)
% forward-backward euler method.
%
% inputs:
%
% dx -- function of t, x, and y
% dy -- function of t, x, and y
% x0 -- initial x
% T -- boundary position
% N -- number of grid points
% tol -- tolerance for convergence
% maxiter -- Maximum number of iterations
%
% outputs: 
%
% x -- array of x values
% y -- array of y values
% t -- array of t values
% m -- number of steps to convergence
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% make a grid
h = (T+1)/N;
t = 0:h:T;
x = x0*ones(1, length(t));
% initialize y as zeros. We can discuss if this is valid or not later.
y = zeros(1, length(t));
xlast = x*0;
ylast = y*0;

convInd = zeros(1,maxiter);

% computation
m = 1;
while m<maxiter && max([ norm(y-ylast), norm(x-xlast) ]) > tol

    % find x
    xlast = x;
    for i = 1:length(t)-1
        x(i+1) = x(i) + h * dx(t,x(i),ylast(i));
    end

    %find y
    ylast = y;
    for i = length(t):-1:2
        y(i-1) = y(i) - h * dy(t,xlast(i),y(i));
    end

    
    % print convergence
    %fprintf(num2str(max([ norm(y-ylast), norm(x-xlast) ])));
     %   fprintf('\n')
     convInd(m) = max([ norm(y-ylast), norm(x-xlast) ]);
     m = m+1;
end

convInd = convInd(1:m-1);
m = m-1;
    
end

File: fbeuler_gs.cpp
#include <iostream>
#include <armadillo>
#include <functional>

using namespace std;
using namespace arma;


double dx(vec t, double x, double y){
  return -x + (2*y) + pow((x-y),2);
}

double dy(vec t, double x, double y){
  return pow((x-y),2) + y;
}

void fbeuler_jb( double x0, double T, long N, double tol, long maxiter){
  double h = 0.0001; //(T+1)/N;
  vec t = linspace<vec>(0,h,T);
  vec x = x0 * ones<vec>(N);
  vec y = zeros<vec>(N);
  vec xlast = zeros<vec>(N);;
  vec ylast = zeros<vec>(N);;
  vec convInd = zeros<vec>(maxiter);
  long m = 0;

  do {
    xlast = x;
    for (long i = 0; i < N-1; i++){
      x(i+1) = x(i) + h * dx(t,x(i),ylast(i));
    }

    ylast = y;
    for (long i = N-1; i > 0; i--){
      y(i-1) = y(i) - h * dy(t,xlast(i),y(i));
    }

    convInd[m] = max(norm(y-ylast), norm(x-xlast));
    printf("diff: %6.13f  y: %6.13f\n", convInd[m], y(0));

    m++;
  } while ((m < maxiter) && convInd[m-1] > tol);
}

int main(int argc, char** argv)
  {
  double T = 4.1;
  long N = 6000;
  double x0 = -0.01;
  double tol = pow(10, -7);
  long maxiter = 10000;

  fbeuler_jb(x0, T, N, tol, maxiter);

  return 0;
  }
